### ISSUE TITLE
AWS::CloudFormation -- Improve validation to report error file, line number, and attribute

### ISSUE THREAD

**Issue by shentonfreude** [aws-cloudformation/aws-cloudformation-coverage-roadmap#859]
## Instructions for CloudFormation Coverage New Issues Template

Quick Sample Summary:
1. Title -> AWS::Service::Resource-Attribute-Existing Attribute
2. Scope of request -> AWS::Service::ResourceType::TheResource-ExistingAttribute supports A and B today, but not new C feature.
3. Expected behavior -> In Create, it should include this and this, like the console.  Also include tag creation, like the API allows.
4. Test case recommendation (optional) -> it should help me do X given existing Y (if applicable); it should/shouldn't change Z.
5. Links to existing API doc (optional) -> Updated API docs at https://example.com/api-new.html
6. Category tag (optional) -> Compute
7. Any additional context (optional)

Detailed Instructions to create a new issue here: (delete this text before you save your issue, to keep them as short as possible :) ). 

## 3. Expected behavior

I get an error:
```
An error occurred (ValidationError) when calling the ValidateTemplate operation: Template error: variable names in Fn::Sub syntax must contain only alphanumeric characters, underscores, periods, and colons
```
But would like it to suffix with file and line number where the error occurred, e.g.,
```
... at ttt2service-stack.yml:370
Resource.LaunchConfig.Properties.UserData
```

## 4. Suggest specific test cases

I have a master stack, `ttt2.yml` which uses a nested `ttt2service-stack.yml`. Submitting ttt2.yml failed with the error above. It turns out I had bad quoting in the UserData of a LaunchConfig. But this was totally not determinable from the error message. My UserData problem was due to:
```
      UserData:
        Fn::Base64: !Sub |
          #!/bin/bash -xe
              yum update -y aws-cfn-bootstrap

              /opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource LaunchConfig --region ${AWS::Region}

              # Set flags common to all Docker Django commands
          docker_flags="""
          --log-driver=awslogs
          --log-opt awslogs-group=${LogGroupName}
          -e AWS_DEFAULT_REGION=${AWS::Region}
         ...
         """
```

If CloudFormaiton had told me the nested file (on S3) and line number where it had the syntax problem, I could have saved a few hours of trial-and-error.

This is what I expect from a C compiler or Python interpreter. Not just "you did something wrong... somewhere."

## 5. Helpful Links to speed up research and evaluation

I whined about my slog here, and you replied I should file this ticket.
https://twitter.com/Shentonfreude/status/1393222974703083529

## 6. Category (required) - Will help with tagging and be easier to find by other users to +1

5. Management (CloudTrail, Config...)

## 7. Any additional context (optional)
none.

**Comment by PatMyron** [comment]
Agreed. In the meantime, [`cfn-lint`](https://github.com/aws-cloudformation/cfn-python-lint/) should provide this, and the [VSCode extension](https://github.com/aws-cloudformation/aws-cfn-lint-visual-studio-code) will inline these errors along with autocompletion and documentation links
